config/database_test.go

```diff
@@ -0,0 +1,479 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package config_test
+
+import (
+	"bytes"
+	"fmt"
+	"net/url"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/mattermost/mattermost-server/config"
+	"github.com/mattermost/mattermost-server/model"
+)
+
+func setupConfigDatabase(t *testing.T, cfg *model.Config) (string, func()) {
+	t.Helper()
+	os.Clearenv()
+	truncateTables(t)
+
+	cfgData, err := config.MarshalConfig(cfg)
+	require.NoError(t, err)
+
+	db := sqlx.NewDb(mainHelper.SqlSupplier.GetMaster().Db, *mainHelper.Settings.DriverName)
+	err = config.InitializeConfigurationsTable(db)
+	require.NoError(t, err)
+
+	id := model.NewId()
+	_, err = db.NamedExec("INSERT INTO Configurations (Id, Value, CreateAt, Active) VALUES(:Id, :Value, :CreateAt, TRUE)", map[string]interface{}{
+		"Id":       id,
+		"Value":    cfgData,
+		"CreateAt": model.GetMillis(),
+	})
+	require.NoError(t, err)
+
+	return id, func() {
+		truncateTables(t)
+	}
+}
+
+// assertDatabaseEqualsConfig verifies the active in-database configuration equals the given config.
+func assertDatabaseEqualsConfig(t *testing.T, expectedCfg *model.Config) {
+	t.Helper()
+
+	// These fields require special initialization for our tests.
+	expectedCfg = expectedCfg.Clone()
+	expectedCfg.MessageExportSettings.GlobalRelaySettings = &model.GlobalRelayMessageExportSettings{}
+	expectedCfg.PluginSettings.Plugins = make(map[string]map[string]interface{})
+	expectedCfg.PluginSettings.PluginStates = make(map[string]*model.PluginState)
+
+	var actualCfgData []byte
+	db := sqlx.NewDb(mainHelper.SqlSupplier.GetMaster().Db, *mainHelper.Settings.DriverName)
+	err := db.Get(&actualCfgData, "SELECT Value FROM Configurations WHERE Active")
+	require.NoError(t, err)
+
+	actualCfg, _, err := config.UnmarshalConfig(bytes.NewReader(actualCfgData), false)
+	require.Nil(t, err)
+
+	assert.Equal(t, expectedCfg, actualCfg)
+}
+
+// assertDatabaseNotEqualsConfig verifies the in-database configuration does not equal the given config.
+func assertDatabaseNotEqualsConfig(t *testing.T, expectedCfg *model.Config) {
```

Comment:
This and the above can be combined with a helper that returns `expected, actual`.